### ISSUE TITLE
Windows 8 laptops with NVidia GPUs crashing with Monitor Source.

### ISSUE THREAD

**Issue by RyPope** [jp9000/OBS#411]
As outlined in [this forum post](https://obsproject.com/forum/threads/numerous-errors.11216/). Laptops running Windows 8 with NVidia GPUs streaming with monitor source will crash with a driver error.

![Example image](http://snag.gy/QRcwI.jpg)

Log from the forum post, I can post more logs if requested as I can replicate this 100% of the time within a couple minutes of streaming.:

`
21:10:12: Open Broadcaster Software v0.60b - 32bit (´・ω・｀)
21:10:12: -------------------------------
21:10:12: CPU Name: Intel(R) Core(TM) i7-4700MQ CPU @ 2.40GHz
21:10:12: CPU Speed: 2394MHz
21:10:12: Physical Memory:  4095MB Total, 4095MB Free
21:10:12: stepping id: 3, model 60, family 6, type 0, extmodel 1, extfamily 0, HTT 1, logical cores 8, total cores 4
21:10:12: monitor 1: pos={0, 0}, size={1920, 1080}
21:10:12: Windows Version: 6.2 Build 9200 
21:10:12: Aero is Enabled
21:10:12: -------------------------------
21:10:12: OBS Modules:
21:10:12: Base Address     Module
21:10:12: 01290000         OBS.exe
21:10:12: 670C0000         OBSApi.dll
21:10:12: 68840000         DShowPlugin.dll
21:10:12: 68990000         GraphicsCapture.dll
21:10:12: 68820000         NoiseGate.dll
21:10:12: 68800000         PSVPlugin.dll
21:10:12: ------------------------------------------
21:10:12: Adapter 1
21:10:12:   Video Adapter: Intel(R) HD Graphics 4600
21:10:12:   Video Adapter Dedicated Video Memory: 0
21:10:12:   Video Adapter Shared System Memory: 1879048192
21:10:12:   Video Adapter Output 1: pos={0, 0}, size={1920, 1080}, attached=true
21:10:12: ------------------------------------------
21:10:12: Adapter 2
21:10:12:   Video Adapter: NVIDIA GeForce GTX 770M
21:10:12:   Video Adapter Dedicated Video Memory: 3122688000
21:10:12:   Video Adapter Shared System Memory: 1172213760
21:10:12: =====Stream Start: 2014-02-03, 21:10:12===============================================
21:10:12:   Multithreaded optimizations: On
21:10:12:   Base resolution: 1920x1080
21:10:12:   Output resolution: 1280x720
21:10:12: ------------------------------------------
21:10:12: Nvidia optimus detected, second adapter selected, enabling optimus hint and switching to adapter 1
21:10:12: Loading up D3D10 on Intel(R) HD Graphics 4600 (Adapter 1)...
21:10:12: ------------------------------------------
21:10:12: Audio Format: 48000hz
21:10:12: Playback device Default
21:10:12: ------------------------------------------
21:10:12: Using desktop audio input: Speakers (Realtek High Definition Audio)
21:10:12: ------------------------------------------
21:10:12: Using auxilary audio input: Microphone (Realtek High Definition Audio)
21:10:13: ------------------------------------------
21:10:13: Audio Encoding: AAC
21:10:13:     bitrate: 128
21:10:13: Using Monitor Capture
21:10:13: Scene buffering time set to 400
21:10:13: ------------------------------------------
21:10:13: Video Encoding: x264
21:10:13:     fps: 60
21:10:13:     width: 1280, height: 720
21:10:13:     preset: veryfast
21:10:13:     profile: main
21:10:13:     keyint: 120
21:10:13:     CBR: yes
21:10:13:     CFR: yes
21:10:13:     max bitrate: 2000
21:10:13:     buffer size: 1000
21:10:13: ------------------------------------------
21:10:14: Using RTMP service: Twitch / Justin.tv
21:10:14:   Server selection: rtmp://live-jfk.justin.tv/app
21:10:14:   Interface: Killer e2200 Gigabit Ethernet Controller (NDIS 6.30) (ethernet, 1000 mbps)
21:10:15: Completed handshake with rtmp://live-jfk.justin.tv/app in 323 ms.
21:10:16: SO_SNDBUF was at 261360
21:10:16: SO_SNDBUF is now 65536

Error: Texture->Map failed: 0x887a0005 0x887a0020

Your video driver reported an internal error. Please check for possible hardware / driver issues. This error can also occur if you have enabled opencl in x264 custom settings.
`

**Comment by notr1ch** [comment]
This is a driver / GPU problem. Also this bug tracker is no longer in active use.

**Comment by RyPope** [comment]
Has this been looked into and proven to be a driver issue?

**Comment by jp9000** [comment]
Hasn't been proven, but it's the best theory we have at the moment.

The amount of code windows 8+ monitor capture requires is very small; so small in fact that pretty much everything is out of our hands.

Firstly, the win8+ code is incredibly simple, it basically sums up as:
- Create IDXGIOutputDuplication object
- Call IDXGIOutputDuplication::AcquireNextFrame to get a frame if a new frame is available
- Render or copy the texture data
- Call IDXGIOutputDuplication::ReleaseFrame

That is pretty much ALL that there is to windows 8+ monitor capture.  There's really not "a way to mess it up".  You either do those steps or you do not capture the monitor.  The process is so incredibly simple that the idea that we would be doing something wrong at this point would be somewhat absurd.

Second, on top of that, this map error happens in a completely unrelated section of code.

Thirdly, this same situation (map error or crashes when win8+ monitor capture is active) does not seemed to happen on desktop computers; the evidence may be circumstantial, but it's been consistent and at especially a frequency of reports with laptops to the point where I feel safe in saying that the laptop graphic adapter setups, or more accurately their drivers, are most likely to blame.

To explain further, almost all major laptops these days (such as yours) have two adapters.  The intel integrated adapter used for saving laptop power which is used render the general desktop stuff, and an nvidia for performance, used to render games.  This setup is done to maximize battery life.

However, the windows 8+ monitor capture method employs some new techniques that apparently make it so the two devices communicate with each other in order to try to composite a proper capture from both adapters.  In the reports I've seen, in every crash with win8+ monitor capture on laptops, it's always when the two sets of drivers are communicating with each other; the stack trace always goes both in to the nvidia driver interface libraries and intel driver interface libraries.

Fourthly, the crash has always been in the drivers, not our code, and I just do not get these sort of reports on desktops almost ever.  The error comes from a DirectX return value, and doesn't say specifically what's going on other than "an error in the driver occurred".

Now, is this complete proof?  If you've ever worked with devices and drivers, then you'd know that no one can "prove" anything beyond their best educated guess based on the circumstances.

When two companies are forced to work together with competing devices I would assume it probably tends to be a pretty bureaucratic process.  If something strange is happening and the drivers of two separate companies are involved, then for all I know the companies may just whitewash the problem between each other.  Their priority is to make sure games work; capturing is most likely not something that's tested as extensively as gaming is, probably due to the fact that very few programs do capturing of this sort after all, so those two situations combined probably mean that they just take forever to solve those problems in the drivers, and may cause the companies to delay looking at the problem if they're not high priority.  On desktops this isn't an issue because desktops only have to ever worry about one adapter, not two.  Even when there are two devices on desktops, it's not ever set up in this convoluted way to conserve power like it is with laptops,

So to sum it up (TLDR):
1.) The win8+ monitor code is nearly impossible to perform incorrectly
2.) The vast majority of reports are always when it's win8+ monitor capture on laptops
3.) Laptops have two graphics adapters and thus two sets of drivers and the crashes that happen with win8+ monitor capture have always been when they're communicating
4.) The error comes from DirectX itself saying that an unrecoverable error in the driver or directx occurred.

If I could fix the problem, I would have done so a long time ago, I would spend my time doing it instead of writing out these overly-lengthy responses to try to get people to hear us out.

**Comment by matcheydj** [comment]
Some monitors on nVidia GPU laptops have the display (monitor), and even
HDMI, linked hard to the integrated graphics, which presents problems when
using the nVidia GPU for that type of capture. FYI

On Wed, Jan 28, 2015 at 4:23 PM, Jim <notifications@github.com> wrote:

> Hasn't been proven, but it's the best theory we have at the moment.
>
> The amount of code windows 8+ monitor capture requires is very small; so
> small in fact that pretty much everything is out of our hands.
>
> Firstly, the win8+ code is incredibly simple, it basically sums up as:
>
>    - Create IDXGIOutputDuplication object
>    - Call IDXGIOutputDuplication::AcquireNextFrame to get a frame if a
>    new frame is available
>    - Render or copy the texture data
>    - Call IDXGIOutputDuplication::ReleaseFrame
>
> That is pretty much ALL that there is to windows 8+ monitor capture.
> There's really not "a way to mess it up". You either do those steps or you
> do not capture the monitor. The process is so incredibly simple that the
> idea that we would be doing something wrong at this point would be somewhat
> absurd.
>
> Second, on top of that, this map error happens in a completely unrelated
> section of code.
>
> Thirdly, this same situation (map error or crashes when win8+ monitor
> capture is active) does not seemed to happen on desktop computers; the
> evidence may be circumstantial, but it's been consistent and at especially
> a frequency of reports with laptops to the point where I feel safe in
> saying that the laptop graphic adapter setups, or more accurately their
> drivers, are most likely to blame.
>
> To explain further, almost all major laptops these days (such as yours)
> have two adapters. The intel integrated adapter used for saving laptop
> power which is used render the general desktop stuff, and an nvidia for
> performance, used to render games. This setup is done to maximize battery
> life.
>
> However, the windows 8+ monitor capture method employs some new techniques
> that apparently make it so the two devices communicate with each other in
> order to try to composite a proper capture from both adapters. In the
> reports I've seen, in every crash with win8+ monitor capture on laptops,
> it's always when the two sets of drivers are communicating with each other;
> the stack trace always goes both in to the nvidia driver interface
> libraries and intel driver interface libraries.
>
> Fourthly, the crash has always been in the drivers, not our code, and I
> just do not get these sort of reports on desktops almost ever. The error
> comes from a DirectX return value, and doesn't say specifically what's
> going on other than "an error in the driver occurred".
>
> Now, is this complete proof? If you've ever worked with devices and
> drivers, then you'd know that no one can "prove" anything beyond their best
> educated guess based on the circumstances.
>
> When two companies are forced to work together with competing devices I
> would assume it probably tends to be a pretty bureaucratic process. If
> something strange is happening and the drivers of two separate companies
> are involved, then for all I know the companies may just whitewash the
> problem between each other. Their priority is to make sure games work;
> capturing is most likely not something that's tested as extensively as
> gaming is, probably due to the fact that very few programs do capturing of
> this sort after all, so those two situations combined probably mean that
> they just take forever to solve those problems in the drivers, and may
> cause the companies to delay looking at the problem if they're not high
> priority. On desktops this isn't an issue because desktops only have to
> ever worry about one adapter, not two. Even when there are two devices on
> desktops, it's not ever set up in this convoluted way to conserve power
> like it is with laptops,
>
> So to sum it up (TLDR):
> 1.) The win8+ monitor code is nearly impossible to perform incorrectly
> 2.) The vast majority of reports are always when it's win8+ monitor
> capture on laptops
> 3.) Laptops have two graphics adapters and thus two sets of drivers and
> the crashes that happen with win8+ monitor capture have always been when
> they're communicating
> 4.) The error comes from DirectX itself saying that an unrecoverable error
> in the driver or directx occurred.
>
> If I could fix the problem, I would have done so a long time ago, I would
> spend my time doing it instead of writing out these overly-lengthy
> responses to try to get people to hear us out.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jp9000/OBS/issues/411#issuecomment-71929419>.
>



-- 
David Matchey
